### PR TITLE
Split pull request search from issue; extend with filters

### PR DESCRIPTION
**Pull request by kgromov**
Enhancement
- Add pull request search implementation to split from issues - ``GHPullRequestSearchBuilder``
- Extend with most of search parameters according to [documentation](https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests)
- Added to both GHRepository and GitHub client itself - so it's possible to search for pull requests from certain repository or in general for organization:  ``GHRepository#searchPullRequests(GHPullRequestSearchBuilder search)`` and ``GitHub#searchPullRequests`` accordingly. 
__________________
Inspiration for this extension was own project with custom filters grouped differently to what github UI prvides (and a bit more user intuitive). And as it turned out library did not have a lot of search parameters in order to shorten search and response time (like created/merged dates and ranges, labels etc).

### CODE REVIEW

**Comment by kgromov** [comment]
What I'm doing wrong? Since on my local env ``mvn spotless:apply`` build is successful for both java11 & 17:
Screenshot 2023-08-17 002747

**Comment by bitwiseman** [comment]
Did you commit and push any updates?  I just did so.

File: src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+        this.q("is:pr");
+        return super.list();
+    }
+
+    @Override
+    protected String getApiUrl() {
+        return "/search/issues";
+    }
+
+    public enum Sort {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java:582]:
```suggestion
    /**
     * The sort order values.
     */
    public enum Sort {
```

File: src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+     * rejected review
+     *
+     * @return the gh pull request search builder
+     */
+    public GHPullRequestSearchBuilder reviewRejected() {
+        q("review", REJECTED.getStatus());
+        return this;
+    }
+
+    public GHPullRequestSearchBuilder reviewedBy(GHUser user) {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java:492]:
```suggestion
    /**
     * reviewed by user
     *
     * @param user
     *            the user
     * @return the gh pull request search builder
     */
    public GHPullRequestSearchBuilder reviewedBy(GHUser user) {
```

**Comment by kgromov** [comment]
> Did you commit and push any updates? I just did so.

Sorry my bad - didn't notice those changes in javadoc

File: src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+
+import static org.kohsuke.github.GHPullRequestSearchBuilder.ReviewStatus.*;
+
+/**
+ * Search for pull requests by main search terms in order to narrow down search results.
+ *
+ * @author Konstantin Gromov
+ * @see <a href="https://docs.github.com/en/search-github/searching-on-github/searching-issues-and-pull-requests">Search
+ *      issues & PRs</a>
+ */
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java:14]:
```suggestion
 *      issues and PRs</a>
 */
```

File: src/test/java/org/kohsuke/github/GHRepositoryTest.java [hunk 4]
```diff
@@ -1709,4 +1707,44 @@ public void cannotRetrievePermissionMaintainUser() throws IOException {
         GHPermissionType permission = r.getPermission("alecharp");
         assertThat(permission.toString(), is("UNKNOWN"));
     }
+
+    @Test
+    public void testSearchPullRequests() throws Exception {
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GHRepositoryTest.java:1712]:
```suggestion
    /**
     * Test searching for pull requests.
     *
     * @throws IOException
     *             the exception
     */
    @Test
    public void testSearchPullRequests() throws Exception {
```

File: src/test/java/org/kohsuke/github/AppTest.java [hunk 5]
```diff
@@ -1429,6 +1431,46 @@ public void testIssueSearch() throws IOException {
         }
     }
 
+    @Test
+    public void testPullRequestSearch() throws Exception {
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/AppTest.java:1435]:
```suggestion
    /**
     * Test searching for pull requests.
     *
     * @throws IOException
     *             the exception
     */
    @Test
    public void testPullRequestSearch() throws Exception {
```

**Comment by bitwiseman** [comment]
@kgromov  
Looks like there's some test failures now.

**Comment by kgromov** [comment]
> @kgromov Looks like there's some test failures now.

I'll take a look.
New tests were passed in ``-Dtest.github.useProxy`` and `` -Dtest.github.org=false -Dtest.github.takeSnapshot`` modes respectively.
Should I request access to ``hub4j-test-org`` or it's not necessary and snapshots can be taken from own repository?

**Comment by bitwiseman** [comment]
Snapshots can be taken from anywhere, but we prefer the test org for later rerecording if needed.  


It shouldn't cause failures though.

**Comment by bitwiseman** [comment]
Again, you need to check in files created during `useSnapshot`.

**Comment by kgromov** [comment]
> Again, you need to check in files created during `takeSnapshot`.

@bitwiseman can you please trigger build - I've updated snapshots few days ago

**Comment by kgromov** [comment]
> > Again, you need to check in files created during `takeSnapshot`.
> 
> @bitwiseman can you please trigger build - I've updated snapshots few days ago

And again please (hopefully last time) - I missed one obvious part that wasn't specified in the CONTRIBUTING.md 😊

**Comment by bitwiseman** [comment]
Please update the contributing doc with the instructions that were missing. Thanks!

**Comment by kgromov** [comment]
> Please update the contributing doc with the instructions that were missing. Thanks!

Done (with test data fix)

**Comment by kgromov** [comment]
Hi, @bitwiseman.
Following existing pattern I've added `org.kohsuke.github.GHPullRequestSearchBuilder` to exclusion for code coverage (since it's also has setters for complex GH* objects):
```
<exclude>org.kohsuke.github.GHPullRequestReviewBuilder.DraftReviewComment</exclude>
                      <exclude>org.kohsuke.github.GHIssue.PullRequest</exclude>
                      <exclude>org.kohsuke.github.GHCommitSearchBuilder</exclude>
                      <exclude>org.kohsuke.github.GHRepositorySearchBuilder</exclude>
                      <exclude>org.kohsuke.github.GHUserSearchBuilder</exclude>
```

Is it fine or should I cover all methods for new class?

**Comment by bitwiseman** [comment]
I'd rather you not exclude the methods. Flip that back so I can see what isn't covered. I can override the code coverage CI requirement if it makes sense.

**Comment by bitwiseman** [comment]
I'll take a look shortly.

**Comment by kgromov** [comment]
@bitwiseman, can you trigger builds? I guess last commit was fine (I extended code coverage to 91%) - I just forgot, that java.version is 8

**Comment by kgromov** [comment]
Hi, @bitwiseman. Any news when PR can be merged, thanks?

File: src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+        CREATED,
+        /** The updated. */
+        UPDATED,
+        /** The relevance. */
+        RELEVANCE
+
+    }
+
+    static GHPullRequestSearchBuilder from(GHPullRequestSearchBuilder searchBuilder) {
+        GHPullRequestSearchBuilder builder = new GHPullRequestSearchBuilder(searchBuilder.root());
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java:476]: With the way this is used, you could have someone calling a GHRepository search method, but no using that objects `GitHub` connection.  

```suggestion
    static GHPullRequestSearchBuilder from(GitHub root, GHPullRequestSearchBuilder searchBuilder) {
        GHPullRequestSearchBuilder builder = new GHPullRequestSearchBuilder(root);
```

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 7]
```diff
@@ -1701,6 +1701,17 @@ public GHPullRequestQueryBuilder queryPullRequests() {
...
+     * Retrieves pull requests according to search terms.
+     *
+     * @param search
+     *            {@link GHPullRequestSearchBuilder}
+     * @return pull requests as the paged iterable
+     */
+    public PagedSearchIterable<GHPullRequest> searchPullRequests(GHPullRequestSearchBuilder search) {
+        return GHPullRequestSearchBuilder.from(search).repo(this).list();
+    }
+
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHRepository.java:1714]: I'm not sure I understand the use case for `searchPullRequests(GHPullRequestSearchBuilder search)`, but you have a tests covering it (although, see the previous comment about passing `root`).  

I think we need at least some equivalent to `GitHub.searchPullRequests()`.  

```suggestion
    /**
     * Retrieves pull requests according to search terms.
     *
     * @param search
     *            {@link GHPullRequestSearchBuilder}
     * @return pull requests as the paged iterable
     */
    public PagedSearchIterable<GHPullRequest> searchPullRequests(GHPullRequestSearchBuilder search) {
        return GHPullRequestSearchBuilder.from(search).repo(this).list();
    }

    /**
     * Search for pull requests.
     *
     * @return gh pull request search builder
     */
    public GHPullRequestSearchBuilder searchPullRequests() {
        return root().searchPullRequests().repo(this);
    }

```

**Review by bitwiseman — CHANGES_REQUESTED** [review]
Great job on the test coverage!
This is very close. Last couple questions.

File: src/main/java/org/kohsuke/github/GHRepository.java (hunk 7 above)
Comment (kgromov) [reply src/main/java/org/kohsuke/github/GHRepository.java:1714]: > I'm not sure I understand the use case for `searchPullRequests(GHPullRequestSearchBuilder search)`, but you have a tests covering it (although, see the previous comment about passing `root`).
> 
> I think we need at least some equivalent to `GitHub.searchPullRequests()`.

The main idea was to use  `GHPullRequestSearchBuilder` from repository level and from global level (actually what can be done in github UI). That's why it's used in 2 places and from repository level source search is appended with current repository not to mess up results.

So you suggest to modify:
```
  public PagedSearchIterable<GHPullRequest> searchPullRequests(GHPullRequestSearchBuilder search) {
        return GHPullRequestSearchBuilder.from(search).repo(this).list();
    }
```
to smth like:

```
  public GHPullRequestSearchBuilder searchPullRequests() {
        return GHPullRequestSearchBuilder.from(this).list();
    }
```
like 
```
 GHPullRequestQueryBuilder(GHRepository repo) {
        super(repo.root());
        this.repo = repo;
    }
```
So we can have 2 constructors - 1 for global search from Github.class and 1 from GHRepository.
Then `GHPullRequestSearchBuilder#from` is not needed at all.

**Comment by kgromov** [comment]
> Great job on the test coverage! This is very close. Last couple questions.

Updated - please take a look last commit

**Comment by kgromov** [comment]
> > Great job on the test coverage! This is very close. Last couple questions.
> 
> Updated - please take a look last commit

UPD: rolled back changed type for terms. I would personally make it LinkedHashSet:
```
public abstract class GHSearchBuilder<T> extends GHQueryBuilder<T> {

    /** The terms. */
    protected final List<String> terms = new ArrayList<String>();
```

File: src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+    /**
+     * Instantiates a new GH search builder from repository.
+     *
+     * @param repository
+     *            the gh repository
+     */
+    GHPullRequestSearchBuilder(GHRepository repository) {
+        super(repository.root(), PullRequestSearchResult.class);
+        this.repo(repository);
+    }
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHPullRequestSearchBuilder.java:33]:
```suggestion
```

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 9]
```diff
@@ -1701,6 +1701,15 @@ public GHPullRequestQueryBuilder queryPullRequests() {
         return new GHPullRequestQueryBuilder(this);
     }
 
+    /**
+     * Retrieves pull requests according to search terms.
+     *
+     * @return gh pull request search builder for current repository
+     */
+    public GHPullRequestSearchBuilder searchPullRequests() {
+        return new GHPullRequestSearchBuilder(this);
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHRepository.java:1710]:
```suggestion
        return new GHPullRequestSearchBuilder(this.root).repo(this);
```

**Comment by kgromov** [comment]
> > > > Great job on the test coverage! This is very close. Last couple questions.
> > > 
> > > 
> > > Updated - please take a look last commit
> > 
> > 
> > UPD: rolled back changed type for terms. I would personally make it LinkedHashSe to guarantee terms uniqueness (hopefully it's not affected github API but can add some ambiguity for wirewock mappings):
> > ```
> > public abstract class GHSearchBuilder<T> extends GHQueryBuilder<T> {
> > 
> >     /** The terms. */
> >     protected final List<String> terms = new ArrayList<String>();
> > ```
> 
> That seems like a reasonable change, but maybe make that in a separate PR after we land this one?

yes, since it's related to all search builders

File: src/main/java/org/kohsuke/github/GHRepository.java (hunk 9 above)
Comment (kgromov) [reply src/main/java/org/kohsuke/github/GHRepository.java]: Was about to fix today, but you already did it 👍

**Review by bitwiseman — APPROVED** [review]

**Pull request merged**